distributed/worker.py

```diff
@@ -1307,23 +1306,31 @@ def keys(self) -> list[str]:
         return list(self.data)
 
     async def gather(self, who_has: dict[str, list[str]]) -> dict[str, Any]:
+        """Endpoint used by Scheduler.replicate()"""
+        first = True
         who_has = {
             k: [coerce_to_address(addr) for addr in v]
             for k, v in who_has.items()
             if k not in self.data
         }
-        result, missing_keys, missing_workers = await gather_from_workers(
-            who_has=who_has, rpc=self.rpc, who=self.address
+
+        async def refresh_who_has(keys: list[str]) -> Mapping[str, Collection[str]]:
+            nonlocal first
+            if first:
+                first = False
+                return who_has
+            return await retry_operation(self.scheduler.who_has, keys=keys)
+
+        result, missing_keys = await gather_from_workers(
+            keys=who_has,
+            who_has=refresh_who_has,
```

Comment:
As noted above, the interchangeable use of `who_has` is confusing. Renaming the parameter of `gather_from_workers` to `refresh_who_has` might help.